### PR TITLE
Environment can now be determined by passing in a MetaData URL

### PR DESCRIPTION
**Pull request by mbfrahry**

### CODE REVIEW

File: authentication/azure_sp_objectid.go [hunk 1]
```diff
@@ -3,16 +3,27 @@ package authentication
...
+		env := &azure.Environment{}
+		var err error
+
+		if c.MetadataURL != "" {
+			env, err = AzureEnvironmentByName(ctx, c.MetadataURL, c.Environment)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			env, err = DetermineEnvironment(c.Environment)
```
[the remark this answers is not in the document]
Comment (mbfrahry) [reply authentication/azure_sp_objectid.go:23]: A user might not pass in the metadata endpoint. Are you thinking we should default the metadata endpoint to whatever the current public endpoint is?

File: authentication/azure_sp_objectid.go [hunk 2]
```diff
@@ -3,16 +3,27 @@ package authentication
...
 func buildServicePrincipalObjectIDFunc(c *Config) func(ctx context.Context) (string, error) {
 	return func(ctx context.Context) (string, error) {
-		env, err := DetermineEnvironment(c.Environment)
-		if err != nil {
-			return "", err
+		env := &azure.Environment{}
+		var err error
+
+		if c.MetadataURL != "" {
+			env, err = AzureEnvironmentByName(ctx, c.MetadataURL, c.Environment)
```
[the remark this answers is not in the document]
Comment (mbfrahry) [reply authentication/azure_sp_objectid.go:18]: Sure would

File: authentication/azure_sp_objectid.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (katbyte) [reply authentication/azure_sp_objectid.go:23]: i think that is sensible? might want to run it by @tombuildsstuff for a final opinion

File: authentication/environment.go [hunk 3]
```diff
@@ -53,3 +96,111 @@ func normalizeEnvironmentName(input string) string {
...
+			return aEnv, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find environment %q from endpoint %q", environmentName, endpoint)
+}
+
+// IsEnvironmentAzureStack returns whether a specific Azure Environment is an Azure Stack environment
+func IsEnvironmentAzureStack(ctx context.Context, endpoint string, environmentName string) (bool, error) {
+	if _, ok := environmentTranslationMap[environmentName]; ok {
```
Comment (tombuildsstuff) [inline authentication/environment.go:127]: we'll need to normalize this name/lookup (e.g. `Public` vs `public`)

File: authentication/environment.go [hunk 4]
```diff
@@ -53,3 +96,111 @@ func normalizeEnvironmentName(input string) string {
 	}
 	return output
 }
+
+// AzureEnvironmentByName returns a specific Azure Environment from the specified endpoint
+func AzureEnvironmentByNameFromEndpoint(ctx context.Context, endpoint string, environmentName string) (*azure.Environment, error) {
+	if env, ok := environmentTranslationMap[environmentName]; ok {
```
Comment (tombuildsstuff) [inline authentication/environment.go:102]: we'll need to normalize this name/lookup (e.g. `Public` vs `public`)

File: authentication/environment.go [hunk 5]
```diff
@@ -53,3 +96,111 @@ func normalizeEnvironmentName(input string) string {
...
+}
+
+func buildAzureEnvironment(env Environment) (*azure.Environment, error) {
+	aEnv := &azure.Environment{
+		Name:                       env.Name,
+		ResourceManagerEndpoint:    env.ResourceManager,
+		StorageEndpointSuffix:      env.Suffixes.Storage,
+		ActiveDirectoryEndpoint:    env.Authentication.LoginEndpoint,
+		GraphEndpoint:              env.Graph,
+		KeyVaultEndpoint:           env.Suffixes.KeyVaultDns,
```
Comment (tombuildsstuff) [inline authentication/environment.go:183]: this wants to be:

```suggestion
		KeyVaultEndpoint:          fmt.Sprintf("https://%s/", env.Suffixes.KeyVaultDns),
```

File: authentication/environment.go [hunk 6]
```diff
@@ -53,3 +96,111 @@ func normalizeEnvironmentName(input string) string {
...
+		GalleryEndpoint:            env.Gallery,
+		BatchManagementEndpoint:    env.Batch,
+		SQLDatabaseDNSSuffix:       env.Suffixes.SqlServerHostname,
+		KeyVaultDNSSuffix:          fmt.Sprintf("https://%s/", env.Suffixes.KeyVaultDns),
```
Comment (tombuildsstuff) [inline authentication/environment.go:187]: this wants to be:



```suggestion
		KeyVaultDNSSuffix:           env.Suffixes.KeyVaultDns,
```

**Pull request merged**